lib/ldap_fluff/ad_member_service.rb

```diff
@@ -8,10 +8,32 @@ def initialize(ldap, config)
   end
 
   # get a list [] of ldap groups for a given user
-  # in active directory, this means a recursive lookup
+  # try to use msds-memberOfTransitive if it is supported, otherwise do a recursive loop
   def find_user_groups(uid)
-    data = find_user(uid)
-    _groups_from_ldap_data(data.first)
+    user_data = find_user(uid).first
+
+    begin
+      search = @ldap.search(:base => "", :scope => Net::LDAP::SearchScope_BaseObject, :attributes => ['domainFunctionality'])
+      if !search.nil? && !search.first.nil?
+        domainFunctionality = search.first['domainfunctionality'].first
+        if domainFunctionality.to_i >= 6
+          user_dn = user_data[:distinguishedname].first
+          search = @ldap.search(:base => user_dn, :scope => Net::LDAP::SearchScope_BaseObject, :attributes => ['msds-memberOfTransitive'])
+          if !search.nil? && !search.first.nil?
+            get_groups(search.first['msds-memberoftransitive'])
+          else
+            raise "Transitive query failed"
+          end
+        else
+          raise "Transitive properties not supported"
+        end
+      else
+        raise "Domain functionality query failed"
+      end
+    rescue
+      # If any exceptions are raised, fall back to recursive lookup
+      _groups_from_ldap_data(user_data)
+    end
```

Comment:
Generally, it's not the best practice to utilize exception for code path. See, if there was a network failure during the initial two search calls, the code will fallback and perform another call silently only after than it errors out. At minimum, you should probably define your own exception class and rescue it explicitly, maybe even avoiding exceptions altogether - they are also quite slow compared to if-else.

@ezr-ondrej thoughts?